src/DssEmergencySpell.sol

```diff
@@ -0,0 +1,74 @@
+// SPDX-FileCopyrightText: © 2024 Dai Foundation <www.daifoundation.org>
+// SPDX-License-Identifier: AGPL-3.0-or-later
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+pragma solidity ^0.8.16;
+
+interface ChainlogLike {
+    function getAddress(bytes32 key) external view returns (address);
+}
+
+interface DssEmergencySpellLike {
+    function action() external view returns (address);
+    function cast() external;
+    function description() external view returns (string memory);
+    function done() external view returns (bool);
+    function eta() external view returns (uint256);
+    function expiration() external view returns (uint256);
+    function log() external view returns (address);
+    function nextCastTime() external view returns (uint256);
+    function officeHours() external view returns (bool);
+    function pause() external view returns (address);
+    function schedule() external;
+    function sig() external view returns (bytes memory);
+    function tag() external view returns (bytes32);
+}
+
+abstract contract DssEmergencySpell is DssEmergencySpellLike {
+    ChainlogLike internal constant _log = ChainlogLike(0xdA0Ab1e0017DEbCd72Be8599041a2aa3bA7e740F);
+
+    address public constant log = address(_log);
+    uint256 public constant eta = 0;
+    bytes public constant sig = "";
+    uint256 public constant expiration = type(uint256).max;
+    // @notice Office Hours is always `false` for emergency spells.
+    bool public constant officeHours = false;
+    address public immutable action = address(this);
+    bytes32 public immutable tag = keccak256(abi.encodePacked(address(this)));
```

Comment:
Although this emergency spell is not ever scheduled in pause, `tag` should be `extcodehash(address(this))` for consistency. Or maybe bytes32(0) for all emergencies? Point is that using a different hash for `tag` can confuse others/raise questions. If `extcodehash` is used a view function is needed (at deploy time the wrong hash will be returned).

Reply:
Yeah, that's kind of a fair point. The original `DssExec` uses `extcodehash(_action)`, as the actual code being executed is in `DssAction`.

Reply:
Hmmm, actually `extcodehash(address(this))` [is not possible](https://blog.smlxl.io/evm-contract-construction-93c98cc4ca96).

During the contract initialization, there is no code. `extcodehash` will return `keccak256("")`.

That is the reason why regular spells have a separate [`DssAction`](https://github.com/makerdao/dss-exec-lib/blob/master/src/DssAction.sol) contract , so `sig` can be [`immutable` and set in the constructor](https://github.com/makerdao/dss-exec-lib/blob/69b658f35d8618272cd139dfc18c5713caf6b96b/src/DssExec.sol#L73-L75).

We would have to turn `tag` into a getter function instead.

Reply:
Please check now.

Reply:
That's weird. This is actually being tested [here](https://github.com/dewiz-xyz/dss-emergency-spells/blob/50493ac34c39006c79daf005b0a4f06092161f95/src/DssEmergencySpell.t.integration.sol#L60-L67) and I got the impression it was working fine.
Can you spot an issue in the assertion?

Reply:
BTW when running: `assertEq(ech.deployCodeHash(), ech.tag())` I get an error:
```
[FAIL. Reason: assertion failed: 0xc5d2460186f7233c927e7db2dcc703c0e500b653ca82273b7bfad8045d85a470 != 0x847c86ea0f8470eadb17ca1052fa13cca43cf401321e4c199b2740e841fe2254]
```

Reply:
The issue is that it doesn't matter if you extract `address(this)` into a solidity block first. During the contract initialization, the code is not present in the contract, so you cannot calculate its hash.

Reply:
Hmmm, I just noticed that:
```
assertEq(ech.codehash(), address(ech).codehash);
```
also passes.

So we don't need assembly at all.